### PR TITLE
Reland "[ValueTracking] Compute knownbits from known fp classes"

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch relands https://github.com/llvm/llvm-project/pull/86409.

I mistakenly thought that `Known.makeNegative()` clears the sign bit of `Known.Zero`. This patch fixes the assertion failure by explicitly clearing the sign bit.

### CODE REVIEW

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 1]
```diff
@@ -1118,6 +1118,42 @@ static void computeKnownBitsFromOperator(const Operator *I,
...
+
+      if (Result.SignBit) {
+        if (*Result.SignBit)
+          Known.makeNegative();
+        else
+          Known.makeNonNegative();
+      }
+
+      assert(!Known.hasConflict() && "Bits known to be one AND zero?");
+
```
Comment (arsenm) [inline llvm/lib/Analysis/ValueTracking.cpp:1153]: Delete blank this blank line

**Review by arsenm — APPROVED** [review]

**Comment by dyung** [comment]
Hi @dtcxzyw, one of our internal tests started failing and I bisected it back to this change you made. I have put the repro in #92217, can you take a look?

**Comment by dtcxzyw** [comment]
> Hi @dtcxzyw, one of our internal tests started failing and I bisected it back to this change you made. I have put the repro in #92217, can you take a look?

Thank you! I will have a look :)

**Comment by mikaelholmen** [comment]
Hi @dtcxzyw 
The following starts crashing with this patch:
```
opt -passes="instcombine" bbi-95530.ll -o /dev/null
```
It fails with
```
opt: ../lib/Analysis/ValueTracking.cpp:1152: void computeKnownBitsFromOperator(const llvm::Operator *, const llvm::APInt &, llvm::KnownBits &, unsigned int, const llvm::SimplifyQuery &): Assertion `!Known.hasConflict() && "Bits known to be one AND zero?"' failed.
```
[bbi-95530.ll.gz](https://github.com/llvm/llvm-project/files/15330187/bbi-95530.ll.gz)

**Comment by dtcxzyw** [comment]
> Hi @dtcxzyw The following starts crashing with this patch:
> 
> ```
> opt -passes="instcombine" bbi-95530.ll -o /dev/null
> ```
> 
> It fails with
> 
> ```
> opt: ../lib/Analysis/ValueTracking.cpp:1152: void computeKnownBitsFromOperator(const llvm::Operator *, const llvm::APInt &, llvm::KnownBits &, unsigned int, const llvm::SimplifyQuery &): Assertion `!Known.hasConflict() && "Bits known to be one AND zero?"' failed.
> ```
> 
> [bbi-95530.ll.gz](https://github.com/llvm/llvm-project/files/15330187/bbi-95530.ll.gz)

I will take a look. If necessary, feel free to revert this commit :)

**Comment by dtcxzyw** [comment]
Reduced reproducer:

```
define i64 @test(double %x) {
  %cmp = fcmp uno double %x, 0.000000e+00
  br i1 %cmp, label %if.then1, label %if.else

if.then1:
  br i1 %cmp, label %if.else, label %if.then2

if.then2:
  %cast = bitcast double %x to i64
  %and = and i64 %cast, 1
  ret i64 %and

if.else:
  ret i64 0
}
```

**Comment by dtcxzyw** [comment]
@nikic Do you have a plan that treating the use as a poison value when the knownbit of it has conflicts?

**Comment by jayfoad** [comment]
> Reduced reproducer:
> 
> ```
> define i64 @test(double %x) {
>   %cmp = fcmp uno double %x, 0.000000e+00
>   br i1 %cmp, label %if.then1, label %if.else
> 
> if.then1:
>   br i1 %cmp, label %if.else, label %if.then2
> 
> if.then2:
>   %cast = bitcast double %x to i64
>   %and = and i64 %cast, 1
>   ret i64 %and
> 
> if.else:
>   ret i64 0
> }
> ```

So `if.then2` is dead code. It seems perfectly natural to me that you can get conflicting KnownBits info in dead code, and we should handle it instead of asserting. Treating it as poison sounds good to me.

**Comment by nikic** [comment]
> @nikic Do you have a plan that treating the use as a poison value when the knownbit of it has conflicts?

The plan is to remove all of these assertions and fold conflict to poison. Someone just needs to implement it :)

**Comment by mikaelholmen** [comment]
Thanks for the quick fix @dtcxzyw !

**Pull request merged**